presto-manager-agent/src/main/java/com/teradata/prestomanager/agent/CommandExecutor.java

```diff
@@ -61,22 +67,75 @@ private int execute()
         processBuilder.redirectErrorStream(true);
         try {
             Process process = processBuilder.start();
-            new Thread(() -> {
-                try (InputStream processStream = process.getInputStream()) {
-                    LOGGER.info("Output from command: {}\n{}", commandString, new String(ByteStreams.toByteArray(processStream)));
-                }
-                catch (IOException e) {
-                    LOGGER.error("Failed to log the process output", e);
-                }
-            }).start();
-            if (!process.waitFor(timeout, TimeUnit.SECONDS)) {
+            CommandResult commandResult = new CommandResult();
+            ExecutorService executor = newFixedThreadPool(1);
+            Future<String> outputThread = executor.submit(new ExecuteCallable(process));
+            try {
+                commandResult.setOutput(outputThread.get(timeout, SECONDS));
+            }
+            catch (TimeoutException e) {
                 process.destroyForcibly();
                 throw new PrestoManagerException(format("Command timed out: %s", commandString));
             }
-            return process.exitValue();
+            finally {
+                executor.shutdownNow();
+            }
```

Comment:
I think the things to worry about are processStream, process itself and the thread.
ProcesStream is closed automatically using the try-with-resources block.
process is destroyed if timeout occurs
and executor is shutdown in the finally block. I think all the cases are handled.